### DISCUSSION TITLE
is producer thread-safe?

### DISCUSSION THREAD

**Discussion by sfsf9797** [confluentinc/confluent-kafka-go#1144] [Q&A]
`	p, err := kafka.NewProducer(&kafka.ConfigMap{"bootstrap.servers": bootstrapServers})
`

Could I pass p to multiple goroutines which would call p.Produce() ?

**Comment by adilburaksen** [comment]
Yes, the producer is goroutine-safe. You can share a single `*kafka.Producer` across multiple goroutines and call `Produce()` concurrently.

The Go client is a wrapper around librdkafka, which handles thread safety internally. Multiple goroutines calling `p.Produce()` simultaneously is a common and supported pattern:

```go
p, err := kafka.NewProducer(&kafka.ConfigMap{"bootstrap.servers": bootstrapServers})
if err != nil { ... }
defer p.Close()

// Start a goroutine to handle delivery reports
go func() {
    for e := range p.Events() {
        switch ev := e.(type) {
        case *kafka.Message:
            if ev.TopicPartition.Error != nil {
                log.Printf("Delivery failed: %v", ev.TopicPartition.Error)
            }
        }
    }
}()

// Multiple goroutines can safely call Produce()
var wg sync.WaitGroup
for i := 0; i < 10; i++ {
    wg.Add(1)
    go func(i int) {
        defer wg.Done()
        p.Produce(&kafka.Message{
            TopicPartition: kafka.TopicPartition{Topic: &topic, Partition: kafka.PartitionAny},
            Value:          []byte(fmt.Sprintf("message %d", i)),
        }, nil)
    }(i)
}
wg.Wait()
p.Flush(5000)
```

The main thing to watch: `Produce()` can return `ErrQueueFull` if the internal queue is at capacity under high load. Your goroutines should handle that case (retry or backpressure).
